### PR TITLE
[TR] PIM-4182: fix product values normalization when decimals are not allowed

### PR DESCRIPTION
**Pull request by phaseinducer**
| Q                    | A
| -------------------- | ---
| Bug fix?             | Y
| New feature?         | N
| BC breaks?           | N
| CI currently passes? | Y
| Tests pass?          | Y
| Scenarios pass?      | Y
| Checkstyle issues?*  | N
| PMD issues?**        | N
| Changelog updated?   | Y
| Fixed tickets        | PIM-4182
| DB schema updated?   | N
| Migration script?    | N
| Doc PR               |

### CODE REVIEW

File: features/export/export_products.feature [hunk 1]
```diff
@@ -131,3 +131,38 @@ Feature: Export products
     sku;additional_colors;categories;color;cost-EUR;cost-GBP;cost-USD;country_of_manufacture;customer_rating-tablet;datasheet;description-en_GB-tablet;description-en_US-tablet;enabled;family;groups;handmade;image;legend-en_GB;legend-en_US;locale_specific_attribute-en_US;manufacturer;material;name-en_GB;name-en_US;number_in_stock-tablet;price-EUR;price-GBP;price-USD;release_date-tablet;size;thumbnail;washing_temperature;weight
     tshirt-white;;men_2013,men_2014,men_2015;white;;;;usa;;;;;1;tshirts;;;;;;"attribut specifique";american_apparel;cotton;"White t-shirt";"White t-shirt";;10.00;9.00;15.00;;size_M;;;
     """
+
```
Comment (sumbobyboys) [inline features/export/export_products.feature]: You should add a @jira tag with the url of the ticket :)

File: features/export/export_products.feature [hunk 2]
```diff
@@ -131,3 +131,38 @@ Feature: Export products
...
+  Scenario: Export decimal and metric attributes with the correct decimals formatting
```
Comment (sumbobyboys) [inline features/export/export_products.feature]: Can be split in two simple scenario

File: spec/Pim/Bundle/TransformBundle/Normalizer/Flat/MetricNormalizerSpec.php [hunk 3]
```diff
@@ -96,6 +96,36 @@ function it_normalizes_empty_metric_in_one_fields(MetricInterface $metric)
...
+            ->shouldReturn(['weight' => '72.1000', 'weight-unit' => 'KILOGRAM']);
+    }
+    
+    function it_normalizes_metric_with_float_data_with_decimals_not_allowed(MetricInterface $metric)
+    {
+        $metric->getData()->willReturn('72.0000');
+        $metric->getUnit()->willReturn('KILOGRAM');
+        
+        $this
+            ->normalize($metric, null, ['field_name' => 'weight', 'is_decimals_allowed' => false])
```
Comment (sumbobyboys) [inline spec/Pim/Bundle/TransformBundle/Normalizer/Flat/MetricNormalizerSpec.php]: should be in the previous line

File: spec/Pim/Bundle/TransformBundle/Normalizer/Flat/MetricNormalizerSpec.php [hunk 4]
```diff
@@ -96,6 +96,36 @@ function it_normalizes_empty_metric_in_one_fields(MetricInterface $metric)
...
+            ->shouldReturn(['weight' => '72.1000', 'weight-unit' => 'KILOGRAM']);
+    }
+    
+    function it_normalizes_metric_with_float_data_with_decimals_allowed(MetricInterface $metric)
+    {
+        $metric->getData()->willReturn('72.1000');
+        $metric->getUnit()->willReturn('KILOGRAM');
+        
+        $this
+            ->normalize($metric, null, ['field_name' => 'weight', 'is_decimals_allowed' => true])
```
Comment (sumbobyboys) [inline spec/Pim/Bundle/TransformBundle/Normalizer/Flat/MetricNormalizerSpec.php]: should be in the previous line

File: spec/Pim/Bundle/TransformBundle/Normalizer/Flat/MetricNormalizerSpec.php [hunk 5]
```diff
@@ -96,6 +96,36 @@ function it_normalizes_empty_metric_in_one_fields(MetricInterface $metric)
...
             ->shouldReturn(['weight' => '']);
     }
+    
+    function it_normalizes_metric_with_float_data_with_decimals_allowed_by_default(MetricInterface $metric)
+    {
+        $metric->getData()->willReturn('72.1000');
+        $metric->getUnit()->willReturn('KILOGRAM');
+        
+        $this
+            ->normalize($metric, null, ['field_name' => 'weight'])
```
Comment (sumbobyboys) [inline spec/Pim/Bundle/TransformBundle/Normalizer/Flat/MetricNormalizerSpec.php]: should be in the previous line

File: src/Pim/Bundle/TransformBundle/Normalizer/Flat/MetricNormalizer.php [hunk 6]
```diff
@@ -30,19 +30,20 @@ public function supportsNormalization($data, $format = null)
     public function normalize($object, $format = null, array $context = array())
     {
         $context = $this->resolveContext($context);
+        $isDecimalsAllowed = !array_key_exists('is_decimals_allowed', $context) || $context['is_decimals_allowed'];
```
Comment (BitOne) [inline src/Pim/Bundle/TransformBundle/Normalizer/Flat/MetricNormalizer.php]: isDecimalsAllowed is really weird as decimals is plural.
decimalsAllowed is enough ;)

**Comment by phaseinducer** [comment]
- split behat export scenario (decimal + metric)
- fix some other behat scenarios
- renamed "is_decimals_allowed" to "decimals_allowed"

**Comment by nidup** [comment]
GTM

File: src/Pim/Bundle/TransformBundle/Normalizer/Flat/MetricNormalizer.php [hunk 7]
```diff
@@ -30,19 +30,20 @@ public function supportsNormalization($data, $format = null)
     public function normalize($object, $format = null, array $context = array())
     {
         $context = $this->resolveContext($context);
+        $decimalsAllowed = !array_key_exists('decimals_allowed', $context) || $context['decimals_allowed'];
```
Comment (jjanvier) [inline src/Pim/Bundle/TransformBundle/Normalizer/Flat/MetricNormalizer.php]: || true === $context['decimals_allowed'];

I guess

File: src/Pim/Bundle/TransformBundle/Normalizer/Flat/ProductValueNormalizer.php [hunk 8]
```diff
@@ -67,8 +67,9 @@ public function normalize($entity, $format = null, array $context = [])
             $result = [$fieldName => ''];
         } elseif (is_int($data)) {
             $result = [$fieldName => (string) $data];
-        } elseif (is_float($data)) {
-            $result = [$fieldName => sprintf(sprintf('%%.%sF', $this->precision), $data)];
+        } elseif (is_float($data) || 'decimal' === $entity->getAttribute()->getBackendType()) {
```
Comment (jjanvier) [inline src/Pim/Bundle/TransformBundle/Normalizer/Flat/ProductValueNormalizer.php]: if the attribute is a decimal, should we normalize 1 to 1 or to 1.00 ?

if it's to 1.00, you should invert the is_int and is_float lines

Comment (phaseinducer) [inline src/Pim/Bundle/TransformBundle/Normalizer/Flat/ProductValueNormalizer.php]: The thing is, a decimal attribute value will never be 1 but '1.0000' (as a string, that's the difference between decimal and float Doctrine types). In our case we use decimal Doctrine type for decimal attributes.
That's why I rely on backend type to do the fix, but it's quite ugly.
Maybe we should consider rewrite the normalization system to always use backend type, and not the php type of the input.

File: src/Pim/Bundle/TransformBundle/Normalizer/Flat/MetricNormalizer.php (hunk 7 above)
Comment (phaseinducer) [inline src/Pim/Bundle/TransformBundle/Normalizer/Flat/MetricNormalizer.php]: true !

**Pull request merged**